dmypy does not recheck files when they are removed and then re-added.

* Are you reporting a bug, or opening a feature request?
Bug
* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.

In a folder, have two files `test1.py` and `test2.py`. Contents are:

test1.py:
```
x: str = 1
```

test2.py:
```
y: int = 'hello'
```

* What is the actual behavior/output?
```
$ python -m mypy --follow-imports=skip --ignore-missing-imports --cache-fine-grained test1.py
test1.py:1: error: Incompatible types in assignment (expression has type "int", variable has type "str")
$ python -m mypy --follow-imports=skip --ignore-missing-imports --cache-fine-grained test2.py
test2.py:1: error: Incompatible types in assignment (expression has type "str", variable has type "int")
$ dmypy start -- --follow-imports=skip --ignore-missing-imports --use-fine-grained-cache
Daemon started
$ dmypy check test1.py
test1.py:1: error: Incompatible types in assignment (expression has type "int", variable has type "str")
$ dmypy check test1.py
test1.py:1: error: Incompatible types in assignment (expression has type "int", variable has type "str")
$ dmypy check test2.py
test2.py:1: error: Incompatible types in assignment (expression has type "str", variable has type "int")
$ dmypy check test2.py
test2.py:1: error: Incompatible types in assignment (expression has type "str", variable has type "int")
$ dmypy check test1.py
$ dmypy check test2.py
$
```
* What is the behavior/output you expect?
I expect that files are rechecked and errors re-outputted after being removed and then re-added to `dmypy`.
* What are the versions of mypy and Python you are using?
mypy 0.610, python 3.6.3

--- a/mypy/dmypy_server.py
+++ b/mypy/dmypy_server.py
@@ -857,6 +857,20 @@ class Server:
             assert path
             removed.append((source.module, path))
 
+        # Always add modules that were (re-)added, since they may be detected as not changed by
+        # fswatcher (if they were actually not changed), but they may still need to be checked
+        # in case they had errors before they were deleted from sources on previous runs.
+        previous_modules = {source.module for source in self.previous_sources}
+        changed_set = set(changed)
+        changed.extend(
+            [
+                (source.module, source.path)
+                for source in sources
+                if source.module not in previous_modules
+                and (source.module, source.path) not in changed_set
+            ]
+        )
+
         # Find anything that has had its module path change because of added or removed __init__s
         last = {s.path: s.module for s in self.previous_sources}
         for s in sources:

--- a/mypy/dmypy_server.py
+++ b/mypy/dmypy_server.py
@@ -866,7 +866,8 @@ class Server:
             [
                 (source.module, source.path)
                 for source in sources
-                if source.module not in previous_modules
+                if source.path
+                and source.module not in previous_modules
                 and (source.module, source.path) not in changed_set
             ]
         )

--- a/test-data/unit/daemon.test
+++ b/test-data/unit/daemon.test
@@ -62,6 +62,28 @@ Daemon started
 \[mypy]
 files = ./foo.py
 
+[case testDaemonRunMultipleStrict]
+$ dmypy run -- foo.py --strict --follow-imports=error
+Daemon started
+foo.py:1: error: Function is missing a return type annotation
+foo.py:1: note: Use "-> None" if function does not return a value
+Found 1 error in 1 file (checked 1 source file)
+== Return code: 1
+$ dmypy run -- bar.py --strict --follow-imports=error
+bar.py:1: error: Function is missing a return type annotation
+bar.py:1: note: Use "-> None" if function does not return a value
+Found 1 error in 1 file (checked 1 source file)
+== Return code: 1
+$ dmypy run -- foo.py --strict --follow-imports=error
+foo.py:1: error: Function is missing a return type annotation
+foo.py:1: note: Use "-> None" if function does not return a value
+Found 1 error in 1 file (checked 1 source file)
+== Return code: 1
+[file foo.py]
+def f(): pass
+[file bar.py]
+def f(): pass
+
 [case testDaemonRunRestart]
 $ dmypy run -- foo.py --follow-imports=error
 Daemon started

--- a/test-data/unit/fine-grained.test
+++ b/test-data/unit/fine-grained.test
@@ -10340,3 +10340,24 @@ reveal_type(x)
 [out]
 ==
 a.py:3: note: Revealed type is "Union[def (x: builtins.int) -> builtins.int, def (*x: builtins.int) -> builtins.int]"
+
+[case testErrorInReAddedModule]
+# flags: --disallow-untyped-defs --follow-imports=error
+# cmd: mypy a.py
+# cmd2: mypy b.py
+# cmd3: mypy a.py
+
+[file a.py]
+def f(): pass
+[file b.py]
+def f(): pass
+[file unrelated.txt.3]
+[out]
+a.py:1: error: Function is missing a return type annotation
+a.py:1: note: Use "-> None" if function does not return a value
+==
+b.py:1: error: Function is missing a return type annotation
+b.py:1: note: Use "-> None" if function does not return a value
+==
+a.py:1: error: Function is missing a return type annotation
+a.py:1: note: Use "-> None" if function does not return a value

